spec/ics-3-connection-semantics/README.md

```diff
@@ -0,0 +1,401 @@
+
+#### During Handshake
+
+Once a negotiation handshake has begun:
+
+- Only the appropriate handshake datagrams can be executed in order.
+- No third chain can masquerade as one of the two handshaking chains
+
+#### Post-Establishment
+
+Once a negotiation handshake has completed:
+
+- The created connection objects on both chains contain the consensus states specified by the initiating actor.
+- No other connection objects can be maliciously created on other chains by replaying datagrams.
+- The connection should be able to be voluntarily & cleanly closed by either blockchain.
+- The connection should be able to be immediately closed upon discovery of a consensus equivocation.
+
+## Technical Specification
+
+### Data Structures
+
+This ICS defines the `Connection` type:
+
+```golang
+type ConnectionState enum {
+  INIT
+  TRYOPEN
+  TRYCLOSE
+  OPEN
+  CLOSED
+}
+```
+
+```golang
+type Connection struct {
+  state                         ConnectionState
+  counterpartyIdentifier        Identifier
+  clientIdentifier              Identifier
+  counterpartyClientIdentifier  Identifier
+  nextTimeoutHeight             uint64
+}
+```
+
+### Subprotocols
+
+This ICS defines two subprotocols: opening handshake and closing handshake. Header tracking and closing-by-equivocation are defined in ICS 2. Datagrams defined herein are handled as external messages by the IBC relayer module defined in [ICS 26](../spec/ics-26-relayer-module).
+
+![State Machine Diagram](state.png)
+
+#### Opening Handshake
+
+The opening handshake subprotocol serves to initialize consensus states for two chains on each other.
+
+The opening handshake defines four datagrams: *ConnOpenInit*, *ConnOpenTry*, *ConnOpenAck*, and *ConnOpenConfirm*.
+
+A correct protocol execution flows as follows (note that all calls are made through modules per ICS 25):
+
+| Initiator | Datagram          | Chain |
+| --------- | ----------------- | ----- |
+| Actor     | `ConnOpenInit`    | A     |
+| Relayer   | `ConnOpenTry`     | B     |
+| Relayer   | `ConnOpenAck`     | A     |
+| Relayer   | `ConnOpenConfirm` | B     |
+
+At the end of an opening handshake between two chains implementing the subprotocol, the following properties hold:
+- Each chain has each other's correct consensus state as originally specified by the initiating actor.
+- Each chain has knowledge of and has agreed to its identifier on the other chain.
+
+This subprotocol need not be permissioned, modulo anti-spam measures.
+
+*ConnOpenInit* initializes a connection attempt on chain A.
+
+```golang
+type ConnOpenInit struct {
+  identifier                    Identifier
+  desiredCounterpartyIdentifier Identifier
+  clientIdentifier              Identifier
+  counterpartyClientIdentifier  Identifier
+  nextTimeoutHeight             uint64
```

Comment:
Maybe add a few words in the definition section about timeout heights.

Reply:
I wonder if the explanation in #86 should actually become an ICS...